File: 19.Backend Scaling and Performance Engineering - Part-2/code/typescript/example_2.ts
import express from 'express';
import { Pool } from 'pg';
import Redis from 'ioredis';

const app = express();
const db = new Pool();
const rdb = new Redis();

app.get('/health', async (req, res) => {
    const checks: Record<string, string> = {};
    
    try {
        await db.query('SELECT 1');
        checks['database'] = 'ok';
    } catch {
        checks['database'] = 'unreachable';
    }

    try {
        await rdb.ping();
        checks['redis'] = 'ok';
    } catch {
        checks['redis'] = 'unreachable';
    }

    if (Object.values(checks).includes('unreachable')) {
        return res.status(503).json(checks);
    }
    res.json(checks);
});

File: 19.Backend Scaling and Performance Engineering - Part-2/code/rust/example_2.rs
use actix_web::{get, web, HttpResponse, Responder};
use sqlx::PgPool;
use redis::aio::Connection;
use std::collections::HashMap;

#[get("/health")]
async fn health(db: web::Data<PgPool>, redis_con: web::Data<std::sync::Mutex<Connection>>) -> impl Responder {
    let mut checks = HashMap::new();
    
    if sqlx::query("SELECT 1").execute(db.get_ref()).await.is_ok() {
        checks.insert("database", "ok");
    } else {
        checks.insert("database", "unreachable");
    }
    
    // Simulate redis ping Check
    // if redis::cmd("PING").query_async(&mut *redis_con.lock().unwrap()).await.is_ok() { ... }
    
    if checks.values().any(|&v| v != "ok") {
        HttpResponse::ServiceUnavailable().json(&checks)
    } else {
        HttpResponse::Ok().json(&checks)
    }
}
